"Illegal name" for record in unnamed package

In avaje-record-builder-1.0 a `@RecordBuilder` record in the unnamed package generates a builder FQN that is invalid:

```java
@io.avaje.recordbuilder.RecordBuilder
public record UnnamedPackageRecord(int i) {}
```

```
Compilation failure
java.io.UncheckedIOException: javax.annotation.processing.FilerException: Illegal name .UnnamedPackageRecordBuilder
	at io.avaje.recordbuilder.internal.RecordProcessor.readElement(RecordProcessor.java:112)
	at io.avaje.recordbuilder.internal.RecordProcessor.readElement(RecordProcessor.java:89)
	at io.avaje.recordbuilder.internal.RecordProcessor.process(RecordProcessor.java:67)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:1023)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:939)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1267)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1382)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1234)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:916)
	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:317)
	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:176)
	at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:64)
	at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:50)
Caused by: javax.annotation.processing.FilerException: Illegal name .UnnamedPackageRecordBuilder
	at jdk.compiler/com.sun.tools.javac.processing.JavacFiler.checkName(JavacFiler.java:703)
	at jdk.compiler/com.sun.tools.javac.processing.JavacFiler.checkNameAndExistence(JavacFiler.java:723)
	at jdk.compiler/com.sun.tools.javac.processing.JavacFiler.createSourceOrClassFile(JavacFiler.java:498)
	at jdk.compiler/com.sun.tools.javac.processing.JavacFiler.createSourceFile(JavacFiler.java:435)
	at io.avaje.recordbuilder.internal.APContext.createSourceFile(APContext.java:207)
	at io.avaje.recordbuilder.internal.RecordProcessor.readElement(RecordProcessor.java:101)
	... 12 more
```

--- a/README.md
+++ b/README.md
@@ -131,4 +131,5 @@ public class DefaultsBuilder {
   private CustomClass custom = CustomClass.createDefault();
 
 ...the rest of the builder
+}
 ```

--- a/README.md
+++ b/README.md
@@ -39,7 +39,9 @@ public record ArmoredCore(
 ```
 
 The following builder class will be generated for the above:
-```java
+<details>
+    <summary>Generated Class, (click to expand)</summary>
+<pre content="java">
 /**  Builder class for {@link ArmoredCore} */
 public class ArmoredCoreBuilder {
   private String coreName = "Steel Haze";
@@ -106,4 +108,28 @@ public class ArmoredCoreBuilder {
       return this;
   }
 }
+```
+  </pre>
+</details>
+
+## Default Values
+Using `@DefaultValue` we can directly write the code to set the default value in the generated builder. This allows us to directly write a value or use static methods to set the default builder state.
+```java
+@RecordBuilder
+public record Defaults(
+@DefaultValue("List.of(1,2,3)") List<Integer> list,
+@DefaultValue("24") int num,
+@DefaultValue("\"string val\"") String str,
+@DefaultValue("CustomClass.createDefault()") CustomClass custom) {}
+```
+
+This will generate:
+```java
+public class DefaultsBuilder {
+  private List<Integer> list = List.of(1,2,3);
+  private int num = 24;
+  private String str = "string val";
+  private CustomClass custom = CustomClass.createDefault();
+
+...the rest of the builder
 ```

--- a/README.md
+++ b/README.md
@@ -108,7 +108,6 @@ public class ArmoredCoreBuilder {
       return this;
   }
 }
-```
   </pre>
 </details>
 
@@ -117,10 +116,10 @@ Using `@DefaultValue` we can directly write the code to set the default value in
 ```java
 @RecordBuilder
 public record Defaults(
-@DefaultValue("List.of(1,2,3)") List<Integer> list,
-@DefaultValue("24") int num,
-@DefaultValue("\"string val\"") String str,
-@DefaultValue("CustomClass.createDefault()") CustomClass custom) {}
+    @DefaultValue("List.of(1,2,3)") List<Integer> list,
+    @DefaultValue("24") int num,
+    @DefaultValue("\"string val\"") String str,
+    @DefaultValue("CustomClass.createDefault()") CustomClass custom) {}
 ```
 
 This will generate:

--- a/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/ClassBodyBuilder.java
+++ b/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/ClassBodyBuilder.java
@@ -1,6 +1,5 @@
 package io.avaje.recordbuilder.internal;
 
-import static io.avaje.recordbuilder.internal.APContext.elements;
 import static io.avaje.recordbuilder.internal.Templates.classTemplate;
 import static java.util.stream.Collectors.joining;
 
@@ -15,12 +14,11 @@ public class ClassBodyBuilder {
 
   private ClassBodyBuilder() {}
 
-  static String createClassStart(TypeElement type, String typeParams, boolean isImported) {
+  static String createClassStart(
+      TypeElement type, String typeParams, boolean isImported, String packageName) {
 
     final var components = type.getRecordComponents();
-    final var packageName =
-        elements().getPackageOf(type).getQualifiedName().toString()
-            + (isImported ? ".builder" : "");
+
     final var shortName = type.getSimpleName().toString();
     if (type.getEnclosingElement() instanceof TypeElement) {
       isImported = true;

--- a/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/ClassBodyBuilder.java
+++ b/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/ClassBodyBuilder.java
@@ -7,14 +7,13 @@ import static java.util.stream.Collectors.joining;
 import java.text.MessageFormat;
 import java.util.List;
 import java.util.Optional;
-import java.util.stream.Stream;
 
 import javax.lang.model.element.RecordComponentElement;
 import javax.lang.model.element.TypeElement;
-import javax.lang.model.type.TypeKind;
 
-// TODO better name?
 public class ClassBodyBuilder {
+
+  private ClassBodyBuilder() {}
 
   static String createClassStart(TypeElement type, String typeParams, boolean isImported) {
 

--- a/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/ImportPrism.java
+++ b/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/ImportPrism.java
@@ -190,20 +190,6 @@ final class ImportPrism {
     return null;
   }
 
-  private static <T> T getValue(
-      Map<String, AnnotationValue> memberValues,
-      Map<String, AnnotationValue> defaults,
-      String name,
-      Class<T> clazz) {
-    AnnotationValue av = memberValues.get(name);
-    if (av == null) av = defaults.get(name);
-    if (av == null) {
-      return null;
-    }
-    if (clazz.isInstance(av.getValue())) return clazz.cast(av.getValue());
-    return null;
-  }
-
   private static <T> List<T> getArrayValues(
       Map<String, AnnotationValue> memberValues,
       Map<String, AnnotationValue> defaults,

--- a/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordProcessor.java
+++ b/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordProcessor.java
@@ -22,6 +22,7 @@ import javax.annotation.processing.ProcessingEnvironment;
 import javax.annotation.processing.RoundEnvironment;
 import javax.annotation.processing.SupportedAnnotationTypes;
 import javax.lang.model.SourceVersion;
+import javax.lang.model.element.ElementKind;
 import javax.lang.model.element.Name;
 import javax.lang.model.element.RecordComponentElement;
 import javax.lang.model.element.TypeElement;
@@ -59,7 +60,7 @@ public final class RecordProcessor extends AbstractProcessor {
     for (final TypeElement type :
         ElementFilter.typesIn(
             roundEnv.getElementsAnnotatedWith(typeElement(RecordBuilderPrism.PRISM_TYPE)))) {
-      if (type.getRecordComponents().isEmpty()) {
+      if (type.getKind() != ElementKind.RECORD) {
         logError(type, "Builders can only be generated for record classes");
         continue;
       }

--- a/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordProcessor.java
+++ b/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordProcessor.java
@@ -49,7 +49,6 @@ public final class RecordProcessor extends AbstractProcessor {
 
   @Override
   public boolean process(Set<? extends TypeElement> tes, RoundEnvironment roundEnv) {
-
     final var globalTypeInitializers =
         roundEnv.getElementsAnnotatedWith(typeElement(GlobalPrism.PRISM_TYPE)).stream()
             .map(GlobalPrism::getInstanceOn)
@@ -57,14 +56,11 @@ public final class RecordProcessor extends AbstractProcessor {
 
     InitMap.putAll(globalTypeInitializers);
     APContext.setProjectModuleElement(tes, roundEnv);
-    for (final TypeElement type :
-        ElementFilter.typesIn(
-            roundEnv.getElementsAnnotatedWith(typeElement(RecordBuilderPrism.PRISM_TYPE)))) {
+    for (final TypeElement type : ElementFilter.typesIn(roundEnv.getElementsAnnotatedWith(typeElement(RecordBuilderPrism.PRISM_TYPE)))) {
       if (type.getKind() != ElementKind.RECORD) {
         logError(type, "Builders can only be generated for record classes");
         continue;
       }
-
       readElement(type);
     }
 
@@ -77,7 +73,6 @@ public final class RecordProcessor extends AbstractProcessor {
 
     if (roundEnv.processingOver()) {
       try (var reader = getModuleInfoReader()) {
-
         ModuleReader.read(reader);
       } catch (IOException e) {
         // Can't read module, it's whatever
@@ -91,7 +86,6 @@ public final class RecordProcessor extends AbstractProcessor {
   }
 
   private void readElement(TypeElement type, boolean isImported) {
-
     final var components = type.getRecordComponents();
     final var packageElement = elements().getPackageOf(type);
     var unnamed = Utils.isInUnnamedPackage(isImported, packageElement);
@@ -140,7 +134,6 @@ public final class RecordProcessor extends AbstractProcessor {
       }
 
       if (APContext.isAssignable(type.mainType(), "java.util.Collection")) {
-
         String param0ShortType = type.param0().shortType();
         Name simpleName = element.getSimpleName();
         writer.append(
@@ -149,7 +142,6 @@ public final class RecordProcessor extends AbstractProcessor {
       }
 
       if (APContext.isAssignable(type.mainType(), "java.util.Map")) {
-
         String param0ShortType = type.param0().shortType();
         String param1ShortType = type.param1().shortType();
         Name simpleName = element.getSimpleName();

--- a/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordProcessor.java
+++ b/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordProcessor.java
@@ -93,20 +93,25 @@ public final class RecordProcessor extends AbstractProcessor {
   private void readElement(TypeElement type, boolean isImported) {
 
     final var components = type.getRecordComponents();
+    final var packageElement = elements().getPackageOf(type);
+    var unnamed = Utils.isInUnnamedPackage(isImported, packageElement);
     final var packageName =
-        elements().getPackageOf(type).getQualifiedName().toString()
-            + (isImported ? ".builder" : "");
+        unnamed
+            ? ""
+            : packageElement.getQualifiedName().toString() + (isImported ? ".builder" : "");
     final var shortName = type.getSimpleName().toString();
 
     try (var writer =
-        new Append(createSourceFile(packageName + "." + shortName + "Builder").openWriter())) {
+        new Append(
+            createSourceFile((unnamed ? "" : packageName + ".") + shortName + "Builder")
+                .openWriter())) {
 
       var typeParams =
           type.getTypeParameters().stream()
               .map(Object::toString)
               .collect(joining(", "))
               .transform(s -> s.isEmpty() ? s : "<" + s + ">");
-      writer.append(ClassBodyBuilder.createClassStart(type, typeParams, isImported));
+      writer.append(ClassBodyBuilder.createClassStart(type, typeParams, isImported, packageName));
       final var writeGetters = RecordBuilderPrism.getInstanceOn(type).getters();
       methods(writer, typeParams, shortName, components, writeGetters);
     } catch (final IOException e) {

--- a/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordProcessor.java
+++ b/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordProcessor.java
@@ -107,54 +107,72 @@ public final class RecordProcessor extends AbstractProcessor {
               .transform(s -> s.isEmpty() ? s : "<" + s + ">");
       writer.append(ClassBodyBuilder.createClassStart(type, typeParams, isImported, packageName));
       final var writeGetters = RecordBuilderPrism.getInstanceOn(type).getters();
-      methods(writer, typeParams, shortName, components, writeGetters);
+      methods(new WriteContext(writer, typeParams, shortName, components, writeGetters));
     } catch (final IOException e) {
       throw new UncheckedIOException(e);
     }
   }
 
-  private void methods(
-      Append writer,
-      String typeParams,
-      String shortName,
-      List<? extends RecordComponentElement> components,
-      Boolean writeGetters) {
+  private record WriteContext(
+    Append writer,
+    String typeParams,
+    String shortName,
+    List<? extends RecordComponentElement> components,
+    Boolean writeGetters
+  ) {
+    boolean getters() {
+      return Boolean.TRUE.equals(writeGetters);
+    }
 
-    boolean getters = Boolean.TRUE.equals(writeGetters);
+    void append(String content) {
+      writer.append(content);
+    }
+  }
 
-    for (final var element : components) {
+  private void methods(WriteContext ctx) {
+    for (final var element : ctx.components()) {
       final var type = UType.parse(element.asType());
-      writer.append(methodSetter(element.getSimpleName(), type.shortType(), shortName, typeParams));
-      if (getters) {
-        writer.append(
-            methodGetter(
-                element.getSimpleName(),
-                type.shortType().transform(ProcessorUtils::trimAnnotations),
-                shortName));
+      ctx.append(methodSetter(element.getSimpleName(), type.shortType(), ctx.shortName(), ctx.typeParams()));
+      if (ctx.getters()) {
+        writeGetter(ctx, element, type);
       }
-
       if (APContext.isAssignable(type.mainType(), "java.util.Collection")) {
-        String param0ShortType = type.param0().shortType();
-        Name simpleName = element.getSimpleName();
-        writer.append(
-            methodAdd(
-                simpleName.toString(), type.shortType(), shortName, param0ShortType, typeParams));
+        writeAdd(ctx, element, type);
       }
-
       if (APContext.isAssignable(type.mainType(), "java.util.Map")) {
-        String param0ShortType = type.param0().shortType();
-        String param1ShortType = type.param1().shortType();
-        Name simpleName = element.getSimpleName();
-        writer.append(
-            methodPut(
-                simpleName.toString(),
-                type.shortType().transform(ProcessorUtils::trimAnnotations),
-                shortName,
-                param0ShortType,
-                param1ShortType,
-                typeParams));
+        writePut(ctx, element, type);
       }
     }
-    writer.append("}");
+    ctx.append("}");
+  }
+
+  private static void writePut(WriteContext ctx, RecordComponentElement element, UType type) {
+    String param0ShortType = type.param0().shortType();
+    String param1ShortType = type.param1().shortType();
+    Name simpleName = element.getSimpleName();
+    ctx.append(
+      methodPut(
+        simpleName.toString(),
+        type.shortType().transform(ProcessorUtils::trimAnnotations),
+        ctx.shortName(),
+        param0ShortType,
+        param1ShortType,
+        ctx.typeParams()));
+  }
+
+  private static void writeAdd(WriteContext ctx, RecordComponentElement element, UType type) {
+    String param0ShortType = type.param0().shortType();
+    Name simpleName = element.getSimpleName();
+    ctx.append(
+      methodAdd(
+        simpleName.toString(), type.shortType(), ctx.shortName(), param0ShortType, ctx.typeParams()));
+  }
+
+  private static void writeGetter(WriteContext ctx, RecordComponentElement element, UType type) {
+    ctx.append(
+      methodGetter(
+        element.getSimpleName(),
+        type.shortType().transform(ProcessorUtils::trimAnnotations),
+        ctx.shortName()));
   }
 }

--- a/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Templates.java
+++ b/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Templates.java
@@ -19,7 +19,7 @@ public class Templates {
 
     return MessageFormat.format(
         """
-		   package {0};
+		   {0}
 
 		   {1}
 
@@ -55,7 +55,7 @@ public class Templates {
 		         return new {2}{10}({7});
 		     '}'
 		   """,
-        packageName,
+        packageName.isBlank() ? "" : "package " + packageName + ";",
         imports,
         shortName,
         fields,

--- a/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Templates.java
+++ b/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Templates.java
@@ -1,7 +1,6 @@
 package io.avaje.recordbuilder.internal;
 
 import java.text.MessageFormat;
-import java.util.function.Consumer;
 
 public class Templates {
   private Templates() {}
@@ -28,12 +27,12 @@ public class Templates {
 		   @Generated("avaje-record-builder")
 		   public class {2}Builder{8} '{'
 		   {3}
+
 		     private {2}Builder() '{'
 		     '}'
-
-		     private {2}Builder({4}) '{'
-		       {5}
-		     '}'
+		   """
+            + constructor(constructor)
+            + """
 
 		     /**
 		      * Return a new builder with all fields set to default Java values
@@ -67,6 +66,18 @@ public class Templates {
         fullTypeParams,
         fullTypeParams.transform(s -> s.isEmpty() ? " " : " " + s + " "),
         typeParams);
+  }
+
+  private static String constructor(String constructor) {
+
+    return constructor.isBlank()
+        ? ""
+        : """
+
+		     private {2}Builder({4}) '{'
+		       {5}
+		     '}'
+		   """;
   }
 
   static String methodSetter(

--- a/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Utils.java
+++ b/avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Utils.java
@@ -1,6 +1,18 @@
 package io.avaje.recordbuilder.internal;
 
+import javax.lang.model.element.PackageElement;
+import javax.lang.model.util.ElementFilter;
+
 final class Utils {
   private Utils() {}
 
+  static boolean isInUnnamedPackage(boolean isImported, final PackageElement packageElement) {
+    var module = APContext.getProjectModuleElement();
+
+    return packageElement.isUnnamed()
+        || isImported
+            && module.isUnnamed()
+            && ElementFilter.packagesIn(module.getEnclosedElements()).stream()
+                .allMatch(PackageElement::isUnnamed);
+  }
 }

--- a/blackbox-test-records/src/main/java/Unnamed.java
+++ b/blackbox-test-records/src/main/java/Unnamed.java
@@ -1,0 +1,4 @@
+// import io.avaje.recordbuilder.RecordBuilder;
+// delete the module-info and uncomment to test
+// @RecordBuilder(getters = true)
+// public record Unnamed(String line1, String line2, String city, String country) {}

--- a/blackbox-test-records/src/main/java/io/avaje/recordbuilder/defaults/CustomClass.java
+++ b/blackbox-test-records/src/main/java/io/avaje/recordbuilder/defaults/CustomClass.java
@@ -1,0 +1,10 @@
+package io.avaje.recordbuilder.defaults;
+
+public class CustomClass {
+
+  private CustomClass() {}
+
+  public static CustomClass createDefault() {
+    return new CustomClass();
+  }
+}

--- a/blackbox-test-records/src/main/java/io/avaje/recordbuilder/defaults/Defaults.java
+++ b/blackbox-test-records/src/main/java/io/avaje/recordbuilder/defaults/Defaults.java
@@ -1,0 +1,13 @@
+package io.avaje.recordbuilder.defaults;
+
+import java.util.List;
+
+import io.avaje.recordbuilder.DefaultValue;
+import io.avaje.recordbuilder.RecordBuilder;
+
+@RecordBuilder
+public record Defaults(
+    @DefaultValue("List.of(1,2,3)") List<Integer> list,
+    @DefaultValue("24") int num,
+    @DefaultValue("\"string val\"") String str,
+    @DefaultValue("CustomClass.createDefault()") CustomClass custom) {}

--- a/blackbox-test-records/src/main/java/io/avaje/recordbuilder/test/EmptyRecord.java
+++ b/blackbox-test-records/src/main/java/io/avaje/recordbuilder/test/EmptyRecord.java
@@ -1,0 +1,8 @@
+package io.avaje.recordbuilder.test;
+
+import io.avaje.recordbuilder.RecordBuilder;
+
+@RecordBuilder
+public record EmptyRecord() {
+  // nothing to see here
+}

